### DISCUSSION TITLE
Welcome to sfall Discussions!

### DISCUSSION THREAD

**Discussion by phobos2077** [sfall-team/sfall#350] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by FakelsHub** [comment]
Нужно переписать весь код движка в sfall, вот такие у меня наполеоновские планы.

**Reply by NovaRain** [reply]
> 
> 
> Нужно переписать весь код движка в sfall, вот такие у меня наполеоновские планы.

Napoleonic plans? As long as the rewrite is not in pure ASM, I guess it's fine.

**Reply by FakelsHub** [reply]
Napoleon's plans are plans that are difficult and sometimes impossible to implement. 
All these hacks only make it harder to understand the code, even if you rewrite everything in c++, the code will still be difficult to understand, since it is essentially divided into various small parts.

**Reply by phobos2077** [reply]
Тебе просто надо начать контрибьютить в Falltergeist и помочь дотащить его до финиша :)

**Reply by phobos2077** [reply]
Ещё вариант взять либы для чтения ресурсов из Falltergeist, оформить их обратно в нативную либу, сделать C# прослойку и портировать в Unity проект. После чего написать всю игру на нём)) Или на Godot

**Reply by FakelsHub** [reply]
Мне почемуто не нравиться архитектура кода в Falltergeist.
Если я туда пойду то это приведет к тому что я начну причесывать его код в сторону оригинального движка.

**Reply by phobos2077** [reply]
В каком смысле "причесывать в сторону оригинала"? Оригинал на си написан, в процедурном стиле. Там сплошная лапша и говнокод. Уж точно не надо делать как в оригинале...

**Reply by FakelsHub** [reply]
Вполне нормальный процедурный код для СИ языка.
Причесывать, значит придерживаться к структурам данным и к реализациям оригинала.

Без обид, но в Falltergeist, я видел реальные говно-реализации, да так что аж плакать хочется  при их виде. Такое ощущение, что его практикующиеся студенты писали))
Плюс нужно что-то делать с его тормознутостью (или же это так SDL2 сильно тормозит).

**Reply by phobos2077** [reply]
Не спорю что в Falltergeist говна хватает, но там хотя-бы код более менее организован. Нахрена вот придерживаться к "реализации оригинала"? В процедурном стиле никто уже 30 лет не пишет. Чем это будет лучше sfall? Опять лапша для парочки любителей ковыряться в низкоуровневом болоте. Чтобы привлечь людей надо ориентироваться на современные стандарты, чтобы обычные программисты могли быстро въехать в проект и начать контрибьютить.

**Reply by phobos2077** [reply]
Но если конечно ты хочешь сильно писать на голом си, без классов и неймспейсов - как говорится, флаг в руки. Никто тебя не остановит. Но зачем усложнять задачу себе и снижать шансы проекта на успех в будущем? Чтобы перед глазами был код оригинала ты и так можешь декомпилировать его в Си со всеми процедурами (а благодаря трудам Crafty и co - ещё и с кучей описанных структур). Но использовать чисто чтобы подсмотреть как оно было, для справки так сказать. Повторять это точно не нужно...

**Reply by phobos2077** [reply]
Вообще я думаю тут главное создать полностью рабочий движок - чтобы можно было пройти игру на нём. Одному или с чьей-то помощью, но довести до играбельного состояния, чтобы архитектура проекта была не просто создана (как в Falltergeist) но и проверена на жизнеспособность реализацией большей части функционала. И после этого уже не будет места спорам на тему какой язык или архитектура лучше - переделывать будет уже поздно и народ потянется так или иначе. Другое дело что какой народ.. тех кто любит писать на Си - единицы. Мододелам в принципе насрать на чём написан движок, они будут ковыряться и делать свои мелкие правки. Но вот потенциальных крупных контрибьюторов в движок ты отпугнёшь таким нетрадиционным выбором языка.

**Reply by phobos2077** [reply]
А вообще если уж писать с нуля, я бы выбрал более соверменный язык типа Rust. Смысл римейка движка в том чтобы как-бы увековечить игру - выдернуть её из "того света". Логично использовать для этого более жизнеспособные технологии.

**Reply by FakelsHub** [reply]
Вобще я люблю с#, и я не говорю что я буду писать на си.

**Reply by FakelsHub** [reply]
> бы выбрал более соверменный язык типа Rust

Уже как-бы есть такой и он похоже загнулся как Falltergeist
Rust это же вроде более процедурный,  хотя я в нем не бэ-мэ.
https://github.com/pingw33n/vault13

**Reply by phobos2077** [reply]
> Rust это же вроде более процедурный, хотя я в нем не бэ-мэ.

Я тоже не бэ-мэ но некоторые коллеги очень хорошо о нём отзывались. По сути это современная замена С++. Правда плюсы он может заменить только для новых проектов, там где можно начать всё с чистого листа. Поэтому плюсы никуда не денутся в ближайшие годы, но и писать на них что-то перспективное я бы не стал.

> https://github.com/pingw33n/vault13

Думаю не язык виноват в том что проект загнулся ))

**Comment by FakelsHub** [comment]
By the way, with the help of reengineering, and some parts of the source code found, the engine for the old game [DIABLO 1 ](https://github.com/diasurgical/devilution) was restored

**Reply by NovaRain** [reply]
I guess the similar case for Fallout 2 would be next to impossible. Devilution was lucky because Diablo 1 has a PlayStation port and thanks to shoddy data control of Sony.
Ideally Bethesda could release the source code they got from Interplay, but Bethesda has either to remove the code of integrated Miles Sound System by RAD Game Tools or makes a new agreement with RAD, because the original license agreement with Interplay is they can use the code of MSS, but they cannot redistribute the source.

In a perfect world we should already have many source ports that people can play with the game files and enjoy highly flexible modding like making the game to a Jagged Alliance clone, not being stuck at cramming more and more hacks with DLL injection.

**Comment by phobos2077** [comment]
Вообще-то тут можно создавать отдельные темы для обсуждения. Эта тема - дефолтная которую гитхаб создал, чисто посмотреть как эта штука работает.

**Comment by Lexx2k** [comment]
Isn't it easier to just open issue tickets for each issue that needs discussion?

**Reply by phobos2077** [reply]
Issues are things that needs to be fixed/implemented. A lot of times they are used for endless long discussions. Also I don't think you can do thread-replies in Issues yet.
